design/ceph-deprecate-old-osd-types.md

```diff
@@ -0,0 +1,212 @@
+1. Identify OSD features/code-paths that should be deprecated due to having become outmoded in some form
+2. Propose a new architecture for Rook's OSD code that is compatible with the planned config updates
+and that is extensible for the future.
+3. Provide a timeline for deprecating outmoded, and provide users tools for transitioning their
+   deployments which use old methods to begin using the new methods.
+
+
+## Deprecated features and code paths
+This design proposes to deprecate ...
+- the use of directories as storage targets
+- support for the Filestore backing store
+- support for devices not provisioned with `ceph-volume`
+
+
+## New OSD operator architecture
+The new OSD operator design focuses only on devices provisioned with Bluestore using `ceph-volume`.
+Some design thought is given to allowing space for support of post-Bluestore backing stores.
+
+### Future security concerns
+While planning for future extensibility of OSDs, a notable future development will be attempts to
+reduce the privilege requirements and system access which Rook's OSD-related pods will have. While
+the specific details of how this will be achieved are out of scope for this design, some details are
+known currently which can help influence this design, making it flexible for future developments.
+
+The most critical design choice which can help reduce pod permissions is to give pods access to only
+the host storage that is required for a given operation. This may mean mounting only certain paths
+and/or devices into pods, and it may mean mounting paths/devices in read-only mode when possible.
+
+Reasonable attempts to secure Rook during execution of this design can be made, but it is not in the
+scope of this design that many hours be spent focusing on security initially. Where security is
+mentioned in this design, a reasonable product-of-opportunity attempt may be possible, but
+completion of this design is not predicated on completion of security concerns.
+
+### Operator knowledge
+Currently, the config updates are limited because the Rook operator has limited knowledge of the
+hardware being configured on hosts. The updated config plan calls for a Ceph config file that is
+minimal and standard for all daemons; daemons are individually configured using command line flags.
+In order to properly set command line flags for object storage daemons, the operator must have
+knowledge of the disks which an OSD will use. The operator must configure, for example, an OSD to
+use one disk as the main storage, another disk as wal storage, and another disk as db storage.
+Mechanisms are not currently in place to report the needed disk information to the operator.
+
+### Proposed design
+The existing OSD operator design uses a two-stage process. The first stage is preparation of OSDs on
+a per-host basis, and the second stage is running individual OSDs. The design proposed here
+separates OSD tasks into more stages.
+
+#### Stage 1: Host device inspection
```

Comment:
Are you saying the osd prepare job will be broken into stages 1-3, each with their own jobs? I'm not following how that helps things and it seems more complex. If any of stages 1-3 needs the same privileges as the prepare pod has today (such as access to /dev), it doesn't seem like we're saving anything for future security concerns.

The philosophy around the current prepare pod is:
- The operator passes the job the desired state from the CR
- The prepare pod looks at all the devices, determines what should be provisioned according to the desired state, and then goes ahead and provisions them.
- The list of OSDs created is saved to a configmap
- The operator then reads the configmap and creates the individual OSD deployments.

There will of course be  some refactoring and code cleanup inside the prepare pod as we deprecate the old way of doing things, but I don't understand the need for splitting the job.

Reply:
The specific security thoughts I had in mind were that stage 1 need only have *read-only* access to `/dev`. Stage 2 is just processing, and then stage 3 needs to have *read-write* access to *only* the devices it intends to provision and doesn't need write access to the entirety of `/dev`. Having write access to the entirety of `/dev` is the security hole I hope to fix here.

> The list of OSDs created is saved to a configmap

What configmap is that, by the way? I didn't see anything obvious in my cluster when I ran `kubectl -n rook-ceph get cm`.